sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDB.scala

```diff
@@ -956,13 +956,43 @@ class RocksDB(
       acquiredThreadInfo = newAcquiredThreadInfo
       // Add a listener to always release the lock when the task (if active) completes
       Option(TaskContext.get()).foreach(_.addTaskCompletionListener[Unit] {
-        _ => this.release(StoreTaskCompletionListener)
+        _ => this.releaseForThread(StoreTaskCompletionListener, newAcquiredThreadInfo.threadRef)
```

Comment:
this itself might not be enough right ?

what if you have the following

- load (lock)
- release
- close (lock)
- completion listener release
- close (release)

in which case the last close is still resetting incorrectly ?

Reply:
Discussed offline with @liviazhu-db